File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/path_sum_III/PathSumIII3.java
package main.java.com.jiang.leetcode100.path_sum_III;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) {this.val = val;}
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution3 {
    public int pathSum(TreeNode root, int sum) {
        if(root == null) return 0;
        Map<Integer, Integer> prefix = new HashMap<>();
        prefix.put(0, 1);
        return helper(root, 0, sum, prefix);
    }

    private int helper(TreeNode root, int cur_sum, int sum, Map<Integer, Integer> prefix) {
        if(root == null) return 0;

        cur_sum += root.val;
        int count = prefix.getOrDefault(cur_sum - sum, 0);
        prefix.put(cur_sum, prefix.getOrDefault(cur_sum, 0) + 1);
        count += helper(root.left, cur_sum, sum, prefix) + 
            helper(root.right, cur_sum, sum, prefix);
        prefix.put(cur_sum, prefix.get(cur_sum) - 1);
        return count;
    }
}

public class PathSumIII3 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        int sum = scanner.nextInt();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new  int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        TreeNode root = new TreeNode(nums[0], null, null);
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        for(int i = 1; i < nums.length; i += 2) {
            TreeNode cur = queue.poll();
            if(nums[i] != -1) {
                cur.left = new TreeNode(nums[i], null, null);
                queue.add(cur.left);
            }
            if(nums[i + 1] != -1) {
                cur.right = new TreeNode(nums[i + 1], null, null);
                queue.add(cur.right);
            }
        }

        System.out.println(new Solution3().pathSum(root, sum));
    }
}

File: Leetcode100/leetcode100_python/PathSumIII3.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        def dfs(root, cur_sum, prefix):
            if not root:
                return 0
            cur_sum += root.val
            count = prefix.get(cur_sum - sum, 0)
            prefix[cur_sum] = prefix.get(cur_sum, 0) + 1
            count += dfs(root.left, cur_sum, prefix) + dfs(root.right, cur_sum, prefix)
            prefix[cur_sum] -= 1
            return count
        prefix = {0: 1}
        return dfs(root, 0, prefix)

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    sum = int(input())
    root = TreeNode(nums[0])
    queue = [root]
    for i in range(1, len(nums), 2):
        cur = queue.pop(0)
        if nums[i] != -1:
            cur.left = TreeNode(nums[i])
            queue.append(cur.left)
        if nums[i + 1] != -1:
            cur.right = TreeNode(nums[i + 1])
            queue.append(cur.right)
    solution = Solution()
    print(solution.pathSum(root, sum))